### PR TITLE
RFC: Deprecate CharExt::to_{upper,lower}case

### PR DESCRIPTION
**Pull request by SimonSapin**
[Rendered](https://github.com/SimonSapin/rfcs/blob/remove-char-toloweruppercase/text/0000-remove-char-toloweruppercase.md).

### CODE REVIEW

**Comment by SimonSapin** [comment]
CC @alexcrichton @aturon

**Comment by mahkoh** [comment]
>and replace them with methods on `IteratorExt<Item=char>`.

Why would `IteratorExt` have a special case for characters? Can't you just write a new trait and implement it for `Iterator<Item=char>` types?

>it is tempting to write incorrect code like: [..] Removing these methods removes this temptation

Why is `Some(c).iter().to_lowercase().next().unwrap()` less tempting? I don't see how `Some().iter()` helps you achieve the motivation part of this RFC.

>The `String::to_lowercase` and `String::to_uppercase` convenience methods would stay externally unchanged.

I'm surprised that these return strings and not iterators over chars. This goes against the rest of the stdlib design.

**Comment by SimonSapin** [comment]
> Can't you just write a new trait and implement it for Iterator<Item=char> types?

I’ve added this as an alternative. I don’t have a strong opinion on that particular point.

`Some().iter()` is just a migration for code that already has a `char` is input but yeah, perhaps it shouldn’t be suggested either.

The `String` methods are just convenience wrappers. I wouldn’t be opposed to removing them in favor of e.g. `char_iter.to_lowercase().collect::<String>()`. It can be slightly smarter about pre-allocating space than `size_hint` on an iterator can, but maybe that’s not worth extra API surface.

File: text/0000-remove-char-toloweruppercase.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+Unicode case folding (which is not quite the same as lowercase mapping)
+will be provided in a crate.io library.
+
+
+# Drawbacks
+
+The `CharExt` methods could be convenient in some cases, and removing them might be a loss.
+However, e.g. `(c: char).to_lowercase()` can easily be replaced
+with `Some(c: char).iter().to_lowercase()`.
+
```
Comment (alexcrichton) [inline text/0000-remove-char-toloweruppercase.md]: Another drawback you may wish to add is reduced discoverability. It's arguably easiest to discover these iterators through the `char` type itself, not through the `Iterator` trait.

File: text/0000-remove-char-toloweruppercase.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+# Alternatives
+
+* Status quo.
+* Instead of new methods on `IteratorExt`, have functions that accept `I: Iterator<Item=char>`
+  or a new trait implemented for `I: Iterator<Item=char>`.
+
```
Comment (alexcrichton) [inline text/0000-remove-char-toloweruppercase.md]: Perhaps another alternative would be removing `char::to_{upper,lower}case` altogether? That would entail stabilizing `str::to_{upper,lower}case` in the process, but it's at least an alternative!

**Comment by aturon** [comment]
@SimonSapin Thanks for the RFC!

As a general rule, I've grown pretty wary of APIs that *reduce* the ergonomics of certain patterns to discourage them; much better to *improve* the ergonomics of patterns you want to encourage. In terms of the proposal here, my concrete worries are regressions in ergonomics and discoverability that I'm not sure are outweighed by the motivation of trying to nudge people toward "doing the right thing".

This kind of thing came up at several points in IO reform, where the "right" thing to do was to force people to use `OsString` everywhere, but the ergonomic burden was simply too high and the majority of people ended up assuming unicode anyway. So instead we provide e.g. `args` and `args_os` variants, making both the "simple" and "maximally robust" variants easy to use, and letting people choose what is most appropriate for their application. I think that having the two variants side-by-side serves reasonably well to signal that "there's something interesting going on here, and you might consider the more robust option."

I wonder if we could take a similar approach here, and offer both the simple and full case-changing operations? As I understand it, both are specced as part of Unicode, and we could use suggestive names (like `to_uppercase_simple`) as a signal that fully robust apps should investigate the other choice.

**Comment by SimonSapin** [comment]
I’m not sure about having more than one flavor of case mapping in std, since it’s a choice that (I think) many people don’t want to make. (Eventually there will be a crates.io library with all flavors for those who *do* want a specific behavior.)

**Comment by Florob** [comment]
I'm not too fond of this change. Having these methods on `char` itself seems semantically sensible, and most discoverable. I think the documentation and the fact that an iterator is returned make it sufficiently clear that only looking at the first character is not the intended use.
I concur that doing different variants of case-mapping in std is likely not the way to go. If we went there it would be reasonable to expect that it also handles locale. I think this is better suited for a separate crate.
If people commonly go out of their way to use the presented anti-pattern, that suggests to me that they want a simple `char` to `char` mapping. Personally I probably could be convinced that providing just that and ignoring SpecialCasing.txt altogether, leaving anything else to more full-fledged crates, is the way to go. My preferred implementation would honour the locale independent mappings from SpecialCasing.txt though.

**Comment by oli-obk** [comment]
I'm confused... [`CharExt`][2] doesn't have `to_*case` functions. These are implemented directly on `char`

What I think most people expect is ascii uppercase/lowercase (Even as a german I don't expect `'ä' == 'Ä'.to_lowercase()` to be true). These methods still exist in [`AsciiExt`][1] . Why not simply rename the functions to `to_*case_grapheme` and note in the docs that the user wants `to_*case_ascii` for the simple things.

   [1]: http://doc.rust-lang.org/std/ascii/trait.AsciiExt.html
   [2]: http://doc.rust-lang.org/core/char/trait.CharExt.html

**Comment by oli-obk** [comment]
Another possibility is to have these functions return an enum that either contains a single-code-point char which is yielded with `unwrap()` or a multi-code-point char. Maybe something like `unwrap_grapheme()` on that enum might return an iterator over either the single-code-point char, or over the multi-code-point char

**Comment by SimonSapin** [comment]
> I'm confused... CharExt doesn't have to_*case functions. These are implemented directly on char

It used to. That’s a change more recent than this RFC.

**Comment by SimonSapin** [comment]
> What I think most people expect is ascii uppercase/lowercase (Even as a german I don't expect 'ä' == 'Ä'.to_lowercase() to be true).

I think you expect that not as a german, but as a programmer used to seeing poor Unicode support everywhere. We can do better, and we should.

**Comment by mahkoh** [comment]
>I don't expect 'ä' == 'Ä'.to_lowercase() to be true

Why?

**Comment by oli-obk** [comment]
as @SimonSapin said, I'm used to poor Unicode support. Please note the sentence right afterwards: different explicit functions, without the function that might confuse due to being used to something else.

Or, my second suggestion, which I'd prefer, an enum that makes the difference between single-code-point and multi-code-point case mappings explicit.

**Comment by SimonSapin** [comment]
The ASCII-only case mapping also exist in std: http://doc.rust-lang.org/std/ascii/trait.AsciiExt.html

I don’t think *grapheme* is the right term here. For example, the ligature U+FB01 ﬁ maps to two grapheme clusters F and I.

**Comment by oli-obk** [comment]
Yea, grapheme is rather wrong here. I think the name is quite bikesheddable, the question is whether and how it should be made explicit.

**Comment by archer884** [comment]
This RFC is interesting to me because I was just using this in my afternoon break yesterday (...I'm noticing a pattern here) and because I think I implemented *exactly* the naive (problematic?) code you're talking about: `c.to_lowercase().nth(0).unwrap()`

...But even just on break, I realized intuitively (if intuition can apply to a language you've been toying with for just a few months) that the fact that this is an iterator and I'm unwrapping the first value it spits out means that I'm naively depending on something that can't be considered robust. (Just `.unwrap()` itself is a big heavy hint.)

For my use case, the alternatives mentioned here (`Some(c).iter().etc...`) are definitely not an improvement, and they are effectively the same thing--something that'll blow up as soon as I try to operate on a character that doesn't play nice, right? 

So it's not so much that I'm against this as I just don't (yet?) see the benefit. It may be that I'm simply not understanding the situation at all; Rust has been my (occasionally harsh) introduction to UTF-8.

**Comment by SimonSapin** [comment]
Closing, as the motivation is weak and didn’t find much support.

**Pull request closed without merging**